lumeer-core/src/main/java/io/lumeer/core/facade/SearchFacade.java

```diff
@@ -0,0 +1,174 @@
+ */
+package io.lumeer.core.facade;
+
+import io.lumeer.api.model.Collection;
+import io.lumeer.api.model.Document;
+import io.lumeer.api.model.Query;
+import io.lumeer.api.model.Resource;
+import io.lumeer.api.model.View;
+import io.lumeer.engine.api.data.DataDocument;
+import io.lumeer.storage.api.dao.CollectionDao;
+import io.lumeer.storage.api.dao.DataDao;
+import io.lumeer.storage.api.dao.DocumentDao;
+import io.lumeer.storage.api.dao.ViewDao;
+import io.lumeer.storage.api.query.SearchQuery;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+import javax.enterprise.context.RequestScoped;
+import javax.inject.Inject;
+
+@RequestScoped
+public class SearchFacade extends AbstractFacade {
+
+   @Inject
+   private CollectionDao collectionDao;
+
+   @Inject
+   private DataDao dataDao;
+
+   @Inject
+   private DocumentDao documentDao;
+
+   @Inject
+   private ViewDao viewDao;
+
+   public List<Collection> searchCollections(Query query) {
+      Set<Collection> collections = new HashSet<>();
+
+      if (!isCollectionSelectionOnly(query)) {
+         collections.addAll(getCollectionsByDocumentSearch(query));
+      }
+
+      collections.addAll(getCollectionsByCollectionSearch(query));
+
+      return new ArrayList<>(collections);
+   }
+
+   public List<Document> searchDocuments(Query query) {
+      Map<String, Collection> collections = getCollections(query.getCollectionCodes());
+      Map<String, DataDocument> dataDocuments = getDataDocuments(collections.values(), query);
+
+      return getDocuments(collections, dataDocuments);
+   }
+
+   public List<View> searchViews(Query query) {
+      if (isValidViewSearch(query)) {
+         return getViewsByFulltext(query);
+      }
+      return Collections.emptyList();
+   }
+
+   private boolean isCollectionSelectionOnly(Query query) {
+      return (query.getFulltext() == null || query.getFulltext().isEmpty()) && (query.getFilters() == null || query.getFilters().isEmpty());
+   }
+
+   private List<Collection> getCollectionsByCollectionSearch(Query query) {
+      SearchQuery searchQuery = createSearchQuery(query);
+
+      return collectionDao.getCollections(searchQuery).stream()
+                          .map(this::keepOnlyActualUserRoles)
+                          .collect(Collectors.toList());
+   }
+
+   private List<Collection> getCollectionsByDocumentSearch(Query query) {
+      java.util.Collection<Collection> searchedCollections = getCollections(query.getCollectionCodes()).values();
+      List<Collection> matchedCollections = new ArrayList<>();
+
+      for (Collection collection : searchedCollections) {
+         long documentCount = dataDao.getDataCount(collection.getId(), createSearchQuery(query));
+         if (documentCount > 0) {
```

Comment:
Why don't you use `getDocumentCount` method on Collection?

Reply:
Which method do you mean? Document data is stored in multiple mongo collections (per our collection) and we need to search all of them within the given project.